docs/API_DESIGN.md

```diff
@@ -0,0 +1,658 @@
+/// Set of functionality required to implement to be a compatible DSA verifier.
+INTERFACE Verifier
+  /// Execute the verification of the signature against the payload by using the encapsulated public key material.
+  METHOD verify(payload: []byte, signature: []byte): bool
+```
+
+### `Ed25519Generator`
+
+```pseudocode!
+/// Generates private key material for Ed25519.
+CLASS Ed25519Generator
+  /// Generate the private key material; return Jwk includes private key material.
+  STATIC METHOD generate(): Jwk
+```
+
+### `Ed25519Signer`
+
+```pseudocode!
+/// Implementation of [`Signer`](#signer) for Ed25519.
+CLASS Ed25519Signer IMPLEMENTS Signer
+  CONSTRUCTOR(private_key: Jwk)
+
+  /// Implementation of Signer's sign instance method for Ed25519.
+  METHOD sign(payload: []byte): []byte
+```
+
+### `Ed25519Verifier`
+
+```pseudocode!
+/// Implementation of [`Verifier`](#verifier) for Ed25519.
+CLASS Ed25519Verifier IMPLEMENTS Verifier
+  CONSTRUCTOR(public_key: Jwk)
+
+  /// Implementation of Verifier's dsa_verify instance method for Ed25519.
+  METHOD verify(payload: []byte): bool
+```
+
+# Decentralized Identifier's (DIDs)
+
+## `Did`
+
+```pseudocode!
+/// Representation of a [DID Core Identifier](https://www.w3.org/TR/did-core/#identifiers).
+CLASS Did
+  /// URI represents the complete Decentralized Identifier (DID) URI.
+  /// Spec: https://www.w3.org/TR/did-core/#did-syntax.
+  PUBLIC DATA uri: string
+
+  /// URL represents the DID URI + A network location identifier for a specific resource.
+  /// Spec: https://www.w3.org/TR/did-core/#did-url-syntax.
+  PUBLIC DATA url: string
+
+  /// Method specifies the DID method in the URI, which indicates the underlying method-specific identifier scheme (e.g., jwk, dht, key, etc.).
+  /// Spec: https://www.w3.org/TR/did-core/#method-schemes.
+  PUBLIC DATA method: string
+
+  /// ID is the method-specific identifier in the DID URI.
+  /// Spec: https://www.w3.org/TR/did-core/#method-specific-id.
+  PUBLIC DATA id: string
+
+  /// Params is a map containing optional parameters present in the DID URI. These parameters are method-specific.
+  /// Spec: https://www.w3.org/TR/did-core/#did-parameters.
+  PUBLIC DATA params: Map<string, string>?
+
+  /// Path is an optional path component in the DID URI.
+  /// Spec: https://www.w3.org/TR/did-core/#path.
+  PUBLIC DATA path: string?
+
+  /// Query is an optional query component in the DID URI, used to express a request for a specific representation or resource related to the DID.
+  /// Spec: https://www.w3.org/TR/did-core/#query.
+  PUBLIC DATA query: string?
+
+  /// Fragment is an optional fragment component in the DID URI, used to reference a specific part of a DID document.
+  /// Spec: https://www.w3.org/TR/did-core/#fragment.
+  PUBLIC DATA fragment: string?
+
+  CONSTRUCTOR(uri: string)
+```
+
+### Example: Instantiate from a `did:dht`
+
+```pseudocode!
+uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
+did = new Did(uri)
```

Comment:
what can i do with this `Did`? usually i create a bearerdid so i can sign over something (i.e. a message)

Reply:
it just turns the did string to an object?

Reply:
this parses the DID URI into an object representing the Identifier from [the DID Core spec](https://www.w3.org/TR/did-core/#identifier), so it's use cases would be, for example, dereferencing wherein a framgment is needed to be parsed out. It's functionally equivalent to the concept of a `Url.parse(http_endpoint)` which many languages natively support, but specific to DID URI's